### ISSUE TITLE
Documentation of `ansible.builtin.command` -module is incorrect and functioning arguably wrong

### ISSUE THREAD

**Issue by hmontone** [ansible/ansible#84190]
### Summary

Documentation says in synopsis:

> The command(s) will not be processed through the shell, so variables like $HOSTNAME and operations like "*", "<", ">", "|", ";" and "&" will not work. Use the ansible.builtin.shell module if you need these features.

, but that is not at all how it works by default. There is also an argument `expand_argument_vars`:

> Expands the arguments that are variables, for example $HOME will be expanded before being passed to the command to run.

that is `true` by default.

This is not simply a documentation error IMO. This module _should_ work like synopsis says, nothing should be expanded since it's not executed in shell. If one needs something to be expanded they should use `ansible.builtin.shell` module, period.

Furthermore, variable expansion works incorrectly. If I'm not mistaken it uses Python function `os.path.expandvars` (emphasis is mine):

> Return the argument with environment variables expanded. Substrings of the form $name or ${name} are replaced by the value of environment variable name. **Malformed variable names and references to non-existing variables are left unchanged.**

That behaviour is significantly different from shell expansion since shell expands non-existing variable to empty string. Note that even single quotes around dollar-string in `cmd`, or use of `argv` won't change the behaviour, existing environment variables are expanded in any case if `expand_argument_vars` is not explicitly set to `false`. This, of course, is a problem in Python, not in Ansible. But Ansible uses it and it makes the working of the module very unexpected.

### Issue Type

Bug Report

### Component Name

ansible.builtin.command

### Ansible Version

```console
$ ansible --version
ansible [core 2.17.5]
  config file = None
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.12/site-packages/ansible
  ansible collection location = /home/user/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.12.7 (main, Oct  1 2024, 11:15:50) [GCC 14.2.1 20240910] (/usr/bin/python)
  jinja version = 3.1.4
  libyaml = True
```

### Configuration

```console
$ ansible-config dump --only-changed -t all
CONFIG_FILE() = None
EDITOR(env: EDITOR) = gvim --remote-silent
```

### OS / Environment

$ cat /etc/os-release
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo

### Steps to Reproduce

```yaml
- hosts: localhost
  tasks:
    - ansible.builtin.command:
        cmd: echo $non_existing_environment_variable
    - ansible.builtin.command:
        cmd: echo $_
    - ansible.builtin.command:
        cmd: echo '$_'
    - ansible.builtin.command:
        argv:
          - echo
          - $_
```

### Expected Results

The result for each of the tasks in playbook should contain unexpanded dollar string in stdout.

### Actual Results

```console
ansible-playbook [core 2.17.5]
  config file = None
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.12/site-packages/ansible
  ansible collection location = /home/user/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible-playbook
  python version = 3.12.7 (main, Oct  1 2024, 11:15:50) [GCC 14.2.1 20240910] (/usr/bin/python)
  jinja version = 3.1.4
  libyaml = True
No config file found; using defaults
setting up inventory plugins
Loading collection ansible.builtin from 
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
yaml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
ini declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
toml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python3.12/site-packages/ansible/plugins/callback/default.py
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: main.yml *************************************************************
Positional arguments: main.yml
verbosity: 4
connection: ssh
become_method: sudo
tags: ('all',)
inventory: ('/etc/ansible/hosts',)
forks: 5
1 plays in main.yml

PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
task path: /home/user/main.yml:1
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: user
<127.0.0.1> EXEC /bin/sh -c 'echo ~user && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user/.ansible/tmp `"&& mkdir "` echo /home/user/.ansible/tmp/ansible-tmp-1730189338.3760297-18957-264357471539948 `" && echo ansible-tmp-1730189338.3760297-18957-264357471539948="` echo /home/user/.ansible/tmp/ansible-tmp-1730189338.3760297-18957-264357471539948 `" ) && sleep 0'
Using module file /usr/lib/python3.12/site-packages/ansible/modules/setup.py
<127.0.0.1> PUT /home/user/.ansible/tmp/ansible-local-189527x4_92tf/tmpubbnxc28 TO /home/user/.ansible/tmp/ansible-tmp-1730189338.3760297-18957-264357471539948/AnsiballZ_setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/user/.ansible/tmp/ansible-tmp-1730189338.3760297-18957-264357471539948/ /home/user/.ansible/tmp/ansible-tmp-1730189338.3760297-18957-264357471539948/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/user/.ansible/tmp/ansible-tmp-1730189338.3760297-18957-264357471539948/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/user/.ansible/tmp/ansible-tmp-1730189338.3760297-18957-264357471539948/ > /dev/null 2>&1 && sleep 0'
ok: [localhost]

TASK [ansible.builtin.command] *************************************************
task path: /home/user/main.yml:3
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: user
<127.0.0.1> EXEC /bin/sh -c 'echo ~user && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user/.ansible/tmp `"&& mkdir "` echo /home/user/.ansible/tmp/ansible-tmp-1730189339.2904081-19032-272879930046432 `" && echo ansible-tmp-1730189339.2904081-19032-272879930046432="` echo /home/user/.ansible/tmp/ansible-tmp-1730189339.2904081-19032-272879930046432 `" ) && sleep 0'
Using module file /usr/lib/python3.12/site-packages/ansible/modules/command.py
<127.0.0.1> PUT /home/user/.ansible/tmp/ansible-local-189527x4_92tf/tmp56n4p0bw TO /home/user/.ansible/tmp/ansible-tmp-1730189339.2904081-19032-272879930046432/AnsiballZ_command.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/user/.ansible/tmp/ansible-tmp-1730189339.2904081-19032-272879930046432/ /home/user/.ansible/tmp/ansible-tmp-1730189339.2904081-19032-272879930046432/AnsiballZ_command.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/user/.ansible/tmp/ansible-tmp-1730189339.2904081-19032-272879930046432/AnsiballZ_command.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/user/.ansible/tmp/ansible-tmp-1730189339.2904081-19032-272879930046432/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "cmd": [
        "echo",
        "$non_existing_environment_variable"
    ],
    "delta": "0:00:00.002438",
    "end": "2024-10-29 10:08:59.434929",
    "invocation": {
        "module_args": {
            "_raw_params": "echo $non_existing_environment_variable",
            "_uses_shell": false,
            "argv": null,
            "chdir": null,
            "creates": null,
            "executable": null,
            "expand_argument_vars": true,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true
        }
    },
    "msg": "",
    "rc": 0,
    "start": "2024-10-29 10:08:59.432491",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "$non_existing_environment_variable",
    "stdout_lines": [
        "$non_existing_environment_variable"
    ]
}

TASK [ansible.builtin.command] *************************************************
task path: /home/user/main.yml:5
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: user
<127.0.0.1> EXEC /bin/sh -c 'echo ~user && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user/.ansible/tmp `"&& mkdir "` echo /home/user/.ansible/tmp/ansible-tmp-1730189339.4637432-19048-245473223669896 `" && echo ansible-tmp-1730189339.4637432-19048-245473223669896="` echo /home/user/.ansible/tmp/ansible-tmp-1730189339.4637432-19048-245473223669896 `" ) && sleep 0'
Using module file /usr/lib/python3.12/site-packages/ansible/modules/command.py
<127.0.0.1> PUT /home/user/.ansible/tmp/ansible-local-189527x4_92tf/tmpo05feny1 TO /home/user/.ansible/tmp/ansible-tmp-1730189339.4637432-19048-245473223669896/AnsiballZ_command.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/user/.ansible/tmp/ansible-tmp-1730189339.4637432-19048-245473223669896/ /home/user/.ansible/tmp/ansible-tmp-1730189339.4637432-19048-245473223669896/AnsiballZ_command.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/user/.ansible/tmp/ansible-tmp-1730189339.4637432-19048-245473223669896/AnsiballZ_command.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/user/.ansible/tmp/ansible-tmp-1730189339.4637432-19048-245473223669896/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "cmd": [
        "echo",
        "$_"
    ],
    "delta": "0:00:00.002007",
    "end": "2024-10-29 10:08:59.569752",
    "invocation": {
        "module_args": {
            "_raw_params": "echo $_",
            "_uses_shell": false,
            "argv": null,
            "chdir": null,
            "creates": null,
            "executable": null,
            "expand_argument_vars": true,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true
        }
    },
    "msg": "",
    "rc": 0,
    "start": "2024-10-29 10:08:59.567745",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "/usr/bin/python",
    "stdout_lines": [
        "/usr/bin/python"
    ]
}

TASK [ansible.builtin.command] *************************************************
task path: /home/user/main.yml:7
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: user
<127.0.0.1> EXEC /bin/sh -c 'echo ~user && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user/.ansible/tmp `"&& mkdir "` echo /home/user/.ansible/tmp/ansible-tmp-1730189339.5927708-19064-119507082212270 `" && echo ansible-tmp-1730189339.5927708-19064-119507082212270="` echo /home/user/.ansible/tmp/ansible-tmp-1730189339.5927708-19064-119507082212270 `" ) && sleep 0'
Using module file /usr/lib/python3.12/site-packages/ansible/modules/command.py
<127.0.0.1> PUT /home/user/.ansible/tmp/ansible-local-189527x4_92tf/tmpxp99v8l5 TO /home/user/.ansible/tmp/ansible-tmp-1730189339.5927708-19064-119507082212270/AnsiballZ_command.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/user/.ansible/tmp/ansible-tmp-1730189339.5927708-19064-119507082212270/ /home/user/.ansible/tmp/ansible-tmp-1730189339.5927708-19064-119507082212270/AnsiballZ_command.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/user/.ansible/tmp/ansible-tmp-1730189339.5927708-19064-119507082212270/AnsiballZ_command.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/user/.ansible/tmp/ansible-tmp-1730189339.5927708-19064-119507082212270/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "cmd": [
        "echo",
        "$_"
    ],
    "delta": "0:00:00.001937",
    "end": "2024-10-29 10:08:59.692915",
    "invocation": {
        "module_args": {
            "_raw_params": "echo '$_'",
            "_uses_shell": false,
            "argv": null,
            "chdir": null,
            "creates": null,
            "executable": null,
            "expand_argument_vars": true,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true
        }
    },
    "msg": "",
    "rc": 0,
    "start": "2024-10-29 10:08:59.690978",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "/usr/bin/python",
    "stdout_lines": [
        "/usr/bin/python"
    ]
}

TASK [ansible.builtin.command] *************************************************
task path: /home/user/main.yml:9
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: user
<127.0.0.1> EXEC /bin/sh -c 'echo ~user && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/user/.ansible/tmp `"&& mkdir "` echo /home/user/.ansible/tmp/ansible-tmp-1730189339.716268-19080-262099349726860 `" && echo ansible-tmp-1730189339.716268-19080-262099349726860="` echo /home/user/.ansible/tmp/ansible-tmp-1730189339.716268-19080-262099349726860 `" ) && sleep 0'
Using module file /usr/lib/python3.12/site-packages/ansible/modules/command.py
<127.0.0.1> PUT /home/user/.ansible/tmp/ansible-local-189527x4_92tf/tmps0u38h94 TO /home/user/.ansible/tmp/ansible-tmp-1730189339.716268-19080-262099349726860/AnsiballZ_command.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/user/.ansible/tmp/ansible-tmp-1730189339.716268-19080-262099349726860/ /home/user/.ansible/tmp/ansible-tmp-1730189339.716268-19080-262099349726860/AnsiballZ_command.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/user/.ansible/tmp/ansible-tmp-1730189339.716268-19080-262099349726860/AnsiballZ_command.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/user/.ansible/tmp/ansible-tmp-1730189339.716268-19080-262099349726860/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true,
    "cmd": [
        "echo",
        "$_"
    ],
    "delta": "0:00:00.001891",
    "end": "2024-10-29 10:08:59.821039",
    "invocation": {
        "module_args": {
            "_raw_params": null,
            "_uses_shell": false,
            "argv": [
                "echo",
                "$_"
            ],
            "chdir": null,
            "creates": null,
            "executable": null,
            "expand_argument_vars": true,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true
        }
    },
    "msg": "",
    "rc": 0,
    "start": "2024-10-29 10:08:59.819148",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "/usr/bin/python",
    "stdout_lines": [
        "/usr/bin/python"
    ]
}

PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=4    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:

* [`lib/ansible/modules/command.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/command.py)
* [`lib/ansible/plugins/action/command.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/command.py)

If these files are incorrect, please update the `component name` section of the description or use the [component](https://github.com/ansible/ansibotmini#commands) bot command.

**Comment by hmontone** [comment]
It seems that this problem has been noticed several times (would be a miracle if not, really):

- #69273
- #54162
- #80512

**Comment by hmontone** [comment]
https://github.com/ansible/ansible/issues/54162#issuecomment-604704162:

> no need to update, while ticket is open the issue is supposed to persist, leaning to this not being a bug, since the substitution is intended in all the cases discussed.
>
> I have not really seen a use case for the variables not to be expanded, just demonstrations on when they are.

I describe the case where I noticed this.

My task was to implement Openstack password rotation. Since I couldn't find a way to do that using `openstack.cloud` -collection (it seems to contain only administrative module for using any users password, which in turn naturally requires administrator privileges), I decided to use `ansible.builtin.command` -module with Openstack CLI. The process as follows:

1. Generate new random password
2. Write new password to Vault
3. Change password in Openstack

In case of error:

1. If Vault write was successful, roll it back

I only noticed this problem because randomly generated password happened to contain dollar-sign followed by suitable character so that environment variable actually existed. 

I am baffled by the fact that this is "intentional" and therefore "not a bug". The expansion mechanism here is faulty. I can say that, because when you expand dollar-strings in a string into corresponding environment variables it is **always** a shell context, and this mechanism works significantly differently from shell expansion. On the other hand, for expanding paths there is a clear solution: Use shell. That's how it's supposed to work, even outside of Ansible context.

Now, I understand that it's not easily possible to change this as it would break backwards compatibility. But at least this should be clearly pointed out in documentation, since:

1. this is most certainly unexpected behaviour
2. documentation implies completely opposite, and it is not any different if environment variable in wording is changed from `$HOME` to `$HOSTNAME`.
